[DS-2701] REST API

peterdietz opened the pull request:
Refactoring the REST API to use the DSpace Services API refactor. REST now compiles. It would be helpful for another REST expert to review these changes with a fine-tooth comb, as this felt like a massive set of changes, that undoubtedly some changes may have occurred.

Some REST changes are that things might use a UUID instead of their legacyID.

Also, tomcat doesn't load the webapp, so, I haven't verified functionality.

Tomcat says this:
Failed to startup the DSpace Service Manager: failure starting up spring service manager: null
Failure during filter init: Failed to startup the DSpace Service Manager: failure starting up spring service manager: null:java.lang.RuntimeException: Failed to startup the DSpace Service Manager: failure starting up spring service manager: null
21-Aug-2015 18:13:43.851 SEVERE [localhost-startStop-1] org.apache.catalina.core.StandardContext.startInternal Error listenerStart
21-Aug-2015 18:13:43.862 SEVERE [localhost-startStop-1] org.apache.catalina.core.StandardContext.startInternal Context [/rest] startup failed due to previous errors

peterdietz committed:
Add itemService.countItems(Context, Community)

A convenience method to get community item count

peterdietz committed:
partial refactor of DSpace-REST to use DSpace-Services refactor

Breaking changes include some integer ID is now UUID ID.

peterdietz committed:
Remaining refactor of REST API methods to use DSpace Services

REST compiles now. However, DSpace still doesn't boot, so its hard to verify

tdonohue commented:
Hi @KevinVdV, we're assigning this to you for review / inclusion in the Services API branch.  It's ready for review & compiles.  Peter will create any additional improvements/enhancements as separate PRs.

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/BitstreamResource.java:
```diff
@@ -691,26 +686,18 @@ public Response deleteBitstream(@PathParam("bitstream_id") Integer bitstreamId,
-                }
-            }
+            org.dspace.authorize.ResourcePolicy resourcePolicy = resourcePolicyService.find(context, policyId);
+            if(resourcePolicy.getdSpaceObject() == dspaceBitstream && authorizeService.authorizeActionBoolean(context, dspaceBitstream, org.dspace.core.Constants.REMOVE)) {
```
This type of compare won't work as you are comparing objects & not primitive types.

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java:
```diff
@@ -457,57 +450,56 @@ public Bitstream addItemBitstream(@PathParam("item_id") Integer itemId, InputStr
-                    {
-                        if (policy.getResourceID() == dspaceBitstream.getID())
+                    for (org.dspace.authorize.ResourcePolicy policy : bitstreamsPolicies) {
+                        if (policy.getdSpaceObject() == dspaceBitstream)
```
This type of compare won't work as you are comparing objects & not primitive types.

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/CollectionsResource.java:
```diff
@@ -439,14 +445,14 @@ public Response updateCollection(@PathParam("collection_id") Integer collectionI
-            dspaceCollection.setMetadata(org.dspace.content.Collection.INTRODUCTORY_TEXT, collection.getIntroductoryText());
-            dspaceCollection.setMetadata(org.dspace.content.Collection.SHORT_DESCRIPTION, collection.getShortDescription());
-            dspaceCollection.setMetadata(org.dspace.content.Collection.SIDEBAR_TEXT, collection.getSidebarText());
-            dspaceCollection.update();
```
This update has been removed ?

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java:
```diff
@@ -488,16 +486,15 @@ public Community createCommunity(Community community, @QueryParam("userIP") Stri
-            dspaceCommunity.setMetadata(org.dspace.content.Community.INTRODUCTORY_TEXT, community.getIntroductoryText());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SHORT_DESCRIPTION, community.getShortDescription());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SIDEBAR_TEXT, community.getSidebarText());
-            dspaceCommunity.update();
```
This update has been removed ?

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java:
```diff
@@ -565,17 +562,15 @@ public Collection addCommunityCollection(@PathParam("community_id") Integer comm
-            dspaceCollection.setMetadata(org.dspace.content.Collection.SIDEBAR_TEXT, collection.getSidebarText());
-            dspaceCollection.setLicense(collection.getLicense());
-            dspaceCollection.update();
-            dspaceCommunity.update();
```
These updates have been removed ?

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java:
```diff
@@ -650,14 +645,12 @@ public Community addCommunityCommunity(@PathParam("community_id") Integer commun
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SHORT_DESCRIPTION, community.getShortDescription());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SIDEBAR_TEXT, community.getSidebarText());
-            dspaceCommunity.update();
-            dspaceParentCommunity.update();
```
These updates have been removed ?

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java:
```diff
@@ -729,12 +722,11 @@ public Response updateCommunity(@PathParam("community_id") Integer communityId,
-            dspaceCommunity.setMetadata(org.dspace.content.Community.INTRODUCTORY_TEXT, community.getIntroductoryText());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SHORT_DESCRIPTION, community.getShortDescription());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SIDEBAR_TEXT, community.getSidebarText());
-            dspaceCommunity.update();
```
This update has been removed ?

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java:
```diff
@@ -379,10 +377,9 @@ public Response addItemMetadata(@PathParam("item_id") Integer itemId, List<org.d
+                    itemService.addMetadata(context, dspaceItem, data[0], data[1], data[2], entry.getLanguage(), entry.getValue());
                 }
             }
-            dspaceItem.update();
```
This update has been removed ?

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java:
```diff
@@ -457,57 +450,56 @@ public Bitstream addItemBitstream(@PathParam("item_id") Integer itemId, InputStr
+                dspaceBitstream.setDescription(context, description);
             }
 
-            dspaceBitstream.update();
```
This update has been removed ?

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java:
```diff
@@ -544,13 +534,12 @@ public Bitstream addItemBitstream(@PathParam("item_id") Integer itemId, InputStr
                         dspacePolicy.setStartDate(date);
                     }
 
-                    dspacePolicy.update();
```
This update has been removed ?

KevinVdV commented on dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java:
```diff
@@ -872,11 +849,11 @@ else if (!AuthorizeServiceImpl.authorizeActionBoolean(context, bitstream, org.ds
+                    if (bit.getBitstream() == bitstream)
                     {
-                        bundle.removeBitstream(bitstream);
+                        bundleService.removeBitstream(context, bundle, bitstream);
```
Removing a bitstream from a bundle while looping will cause ConcurrentModificationException, best to use an iterator & remove the bitstream from the iterator prior to removing from the bundle.

KevinVdV commented:
Did a quick review & was able to find a few things that require fixing. Once there are fixed I would recommend a quick rebase (do NOT merge, rebase is cleaner) on the current DS-2701-service-api branch.
There is a conflict in the "ItemService" interface were 2 different methods are added at the bottom. Both of them should of course be present in the merged result

When this is done I see no issues in merging this pull request (aside from the fact that the retrieval of DSpaceObject still uses identifiers but I am aware that this will be addressed later in separate JIRA tickets). So fix & rebase & I will merge it.

peterdietz committed:
Change token holder double hashmap to a Guava BiMap

- Can't hold an eperson in the set, since hibernate is lazy

peterdietz committed:
REST DSO ID is legacy int, UUID is named uuid

peterdietz committed:
Change REST ID's to string, we can look them up with findByIdOrLegacyId

peterdietz committed:
Add back service.update after altering object to fire events

peterdietz committed:
Compare object by ID

peterdietz committed:
Delete in loops to use iterator to avoid ConcurrentModification

peterdietz commented:
@tdonohue @KevinVdV Sorry for the delay on this, but I have addressed the issues that Kevin has brought up, and I would recommend merging this REST refactor into services.

peterdietz commented on dspace-rest/src/main/java/org/dspace/rest/BitstreamResource.java:
```diff
@@ -542,24 +558,9 @@ public Response updateBitstreamData(@PathParam("bitstream_id") Integer bitstream
-            BitstreamStorageServiceImpl.delete(context, newBitstreamId);
-
-            context.complete();
+            UUID newBitstreamId = bitstreamStorageService.store(context, dspaceBitstream, is);
```
This is a good example of the difference between a UI doing logic the wrong, complicated way, and how services can simplify this.
